File: kotlin/src/main/kotlin/leetcode/72.kt
package leetcode

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class MinDistance {
    fun minDistance(word1: String, word2: String): Int {
        val m = word2.length
        val d = MutableList(m + 1) { it }
        for (ch in word1) {
            var pre = d[0]
            d[0] += 1
            for ((i, c) in word2.withIndex()) {
                pre = d[i + 1].also{
                    d[i + 1] = if (c == ch) pre else 1 + minOf(d[i], pre, d[i + 1])
                }
            }
        }
        return d[m]
    }

    @Test
    fun test1() {
        assertEquals(3, minDistance("horse", "ros"))
        assertEquals(5, minDistance("intention", "execution"))
    }
}

File: golang/leetcode/72.go
package main

func minDistance(word1 string, word2 string) int {
	m := len(word2)
	dp, pre := make([]int, m+1), 0
	for i := range dp {
		dp[i] = i
	}
	for i, a := range word1 {
		dp[0], pre = i+1, dp[0]
		for j, b := range word2 {
			if a == b {
				pre, dp[j+1] = dp[j+1], pre
			} else {
				pre, dp[j+1] = dp[j+1], min(pre, dp[j+1], dp[j])+1
			}
		}
	}
	return dp[m]
}
